pennylane_lightning/core/src/simulators/lightning_kokkos/gates/MatrixGateFunctors.hpp

```diff
@@ -104,15 +110,84 @@ template <class Precision> struct multiQubitOpFunctor {
+    using KokkosComplexVector = Kokkos::View<Kokkos::complex<Precision> *>;
+    using KokkosIntVector = Kokkos::View<std::size_t *>;
+    using ScratchViewComplex =
+        Kokkos::View<Kokkos::complex<Precision> *,
+                     Kokkos::DefaultExecutionSpace::scratch_memory_space,
+                     Kokkos::MemoryTraits<Kokkos::Unmanaged>>;
+    using ScratchViewSizeT =
+        Kokkos::View<std::size_t *,
+                     Kokkos::DefaultExecutionSpace::scratch_memory_space,
+                     Kokkos::MemoryTraits<Kokkos::Unmanaged>>;
+    using MemberType = Kokkos::TeamPolicy<>::member_type;
+
+    KokkosComplexVector arr;
+    KokkosComplexVector matrix;
+    KokkosIntVector indices;
+    KokkosIntVector parity;
+    KokkosIntVector rev_wires;
+    KokkosIntVector rev_wire_shifts;
+    std::size_t dim;
+    std::size_t num_qubits;
+
+    NCMultiQubitOpFunctor(KokkosComplexVector arr_, std::size_t num_qubits_,
+                          const KokkosComplexVector &matrix_,
+                          const std::vector<std::size_t> &controlled_wires_,
+                          const std::vector<bool> &controlled_values_,
+                          const std::vector<std::size_t> &wires_) {
+        dim = exp2(wires_.size());
+        arr = arr_;
+        matrix = matrix_;
+        num_qubits = num_qubits_;
+        const auto &[parity_, rev_wires_] =
+            reverseWires(num_qubits_, wires_, controlled_wires_);
+        parity = parity_;
+        std::vector<std::size_t> indices_ =
+            generateBitPatterns(wires_, num_qubits_);
+        ControlBitPatterns(indices_, num_qubits_, controlled_wires_,
+                           controlled_values_);
+        indices = vector2view(indices_);
+    }
+
+    KOKKOS_INLINE_FUNCTION
+    void operator()(const MemberType &teamMember) const {
+        const std::size_t k = teamMember.league_rank();
+        ScratchViewComplex coeffs_in(teamMember.team_scratch(1), dim);
+        ScratchViewSizeT indices_scratch(teamMember.team_scratch(1), dim);
+        const std::size_t offset = parity_2_offset(parity, k);
+
+        if (!teamMember.team_rank()) {
+            Kokkos::parallel_for(Kokkos::ThreadVectorRange(teamMember, dim),
+                                 [&](std::size_t inner_idx) {
+                                     coeffs_in(inner_idx) =
+                                         arr(indices(inner_idx) + offset);
+                                     indices_scratch(inner_idx) =
+                                         indices(inner_idx);
+                                 });
+        }
+        teamMember.team_barrier();
+        Kokkos::parallel_for(
+            Kokkos::TeamThreadRange(teamMember, dim), [&](std::size_t i) {
+                const auto idx = indices_scratch(i) + offset;
+                arr(idx) = 0.0;
+                const std::size_t base_idx = i * dim;
+
+                for (std::size_t j = 0; j < dim; j++) {
```

Comment:
Looks like the memory access pattern is not aligned? @maliasadi We might come back to this later.

Reply:
Hmm.. what do you mean @multiphaseCFD? We're doing the multiplication over rows of the matrix (dim rows) in parallel. We may want to consider `Kokkos::LayoutLeft` and `Kokkos::LayoutRight` to further improve the access of elements sequentially for each row.